tests/e2e/cmux_test.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// These tests are directly validating etcd connection multiplexing.
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/api/v3/version"
+	clientv2 "go.etcd.io/etcd/client/v2"
+	"go.etcd.io/etcd/server/v3/etcdserver/api/etcdhttp"
+)
+
+func TestConnectionMultiplexing(t *testing.T) {
+	BeforeTest(t)
+	for _, tc := range []struct {
+		name      string
+		serverTLS clientConnType
+	}{
+		{
+			name:      "ServerTLS",
+			serverTLS: clientTLS,
+		},
+		{
+			name:      "ServerNonTLS",
+			serverTLS: clientNonTLS,
+		},
+		{
+			name:      "ServerTLSAndNonTLS",
+			serverTLS: clientTLSAndNonTLS,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			cfg := etcdProcessClusterConfig{clusterSize: 1, clientTLS: tc.serverTLS, enableV2: true}
+			clus, err := newEtcdProcessCluster(t, &cfg)
+			require.NoError(t, err)
+			defer clus.Close()
+
+			var clientScenarios []clientConnType
+			switch tc.serverTLS {
+			case clientTLS:
+				clientScenarios = []clientConnType{clientTLS}
+			case clientNonTLS:
+				clientScenarios = []clientConnType{clientNonTLS}
+			case clientTLSAndNonTLS:
+				clientScenarios = []clientConnType{clientTLS, clientNonTLS}
+			}
+
+			for _, connType := range clientScenarios {
+				name := "ClientNonTLS"
+				if connType == clientTLS {
+					name = "ClientTLS"
+				}
+				t.Run(name, func(t *testing.T) {
+					testConnectionMultiplexing(t, ctx, clus.EndpointsV3()[0], connType)
+				})
+			}
+		})
+	}
+
+}
+
+func testConnectionMultiplexing(t *testing.T, ctx context.Context, endpoint string, connType clientConnType) {
```

Comment:
The `Tests / test (linux-amd64-fmt) (pull_request) ` failed because `ctx` isn't the first parameter.  Curious why we did not see the same error in https://github.com/etcd-io/etcd/pull/15479?

Reply:
Hmm, maybe older version of `revive` was not as picky? or we removed it.